From ba765f7fc7ec17d29f575105ab5d7d3da3012ed5 Mon Sep 17 00:00:00 2001
From: Vincent Abriou <vincent.abriou@st.com>
Date: Fri, 16 Sep 2016 11:48:30 +0200
Subject: [PATCH 53/61] compositor-st: wait for vblank of all sprites

Wait for the reception of *all* sprite vblank before starting to assign
new planes.
Otherwise, we start with some sprites still in use, which causes some
plane glitches on screen.

Change-Id: Ic022a5e279c0dfe75c404d5c598d905958b5b07a
Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>
Reviewed-on: https://gerrit.st.com/58018
Reviewed-by: Vincent ABRIOU <vincent.abriou@st.com>
Tested-by: Vincent ABRIOU <vincent.abriou@st.com>
---
 libweston/compositor-st.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/libweston/compositor-st.c b/libweston/compositor-st.c
index 9e971a7..f56c9a4 100644
--- a/libweston/compositor-st.c
+++ b/libweston/compositor-st.c
@@ -864,7 +864,7 @@ drm_output_repaint(struct weston_output *output_base,
 			s->output = NULL;
 		}
 
-		output->vblank_pending = 1;
+		output->vblank_pending++;
 	}
 
 	return 0;
@@ -972,7 +972,7 @@ vblank_handler(int fd, unsigned int frame, unsigned int sec, unsigned int usec,
 			 WP_PRESENTATION_FEEDBACK_KIND_HW_CLOCK;
 
 	drm_output_update_msc(output, frame);
-	output->vblank_pending = 0;
+	output->vblank_pending--;
 
 	drm_output_release_fb(output, s->current);
 
@@ -985,7 +985,7 @@ vblank_handler(int fd, unsigned int frame, unsigned int sec, unsigned int usec,
 	s->next = NULL;
 	s->vblank_pending = false;
 
-	if (!output->page_flip_pending) {
+	if (!output->vblank_pending && !output->page_flip_pending) {
 		ts.tv_sec = sec;
 		ts.tv_nsec = usec * 1000;
 		weston_output_finish_frame(&output->base, &ts, flags);
-- 
2.7.4

